task6/script.js

```diff
@@ -0,0 +1,508 @@
+   validUntil: new Date('2021-03-25T23:00:00'),
+   discount: '12%',
+   hashTags: ['одежда'],
+   rating: 5.0,
+   reviews :[]
+
+ },
+ ]
+
+ let getAds = (function(skip = 0, top = 10, filterConfig = {}){
+    let adsResult = [];
+    if (Object.keys(filterConfig).length !== 0 && filterConfig.constructor === Object) {
+        let key = Object.keys(filterConfig);
+        let value = filterConfig[key[0]];
+        if (key[0] === 'createdAt' || key[0] === 'validUntil') {
+            adList.forEach(elem => {
+                let timeRequest = String(new Date(value));
+                let time = String(elem[key]);
+                if (time == timeRequest)
+                    resultAds.push(elem);
+            });
+        } else {
+            adList.forEach(elem => {
+                if (elem[key] === value)
+                    resultAds.push(elem);
+            });
+        }
+
+        if (resultAds.length !== 0) {
+            resultAds.sort(function (a, b) {
+                return new Date(a.createdAt) - new Date(b.createdAt);
+            });
+        }
+    } else {
+        resultAds = adList.slice();
+        resultAds.sort(function (a, b) {
+            return new Date(a.createdAt) - new Date(b.createdAt);
+        });
+    }
+
+    if (resultAds.length !== 0) {
+        let border = skip + top;
+        if (border > resultAds.length)
+            resultAds = resultAds.slice(skip, resultAds.length);
+        else
+            resultAds = resultAds.slice(skip, border);
+    }
+    return adsResult;
+});
+
+ let getAd = (function(id){
+    for (let i =0; i<adList.length;++i){
+        if(adList[i].id == id){
+            return adList[i];
+        }
+    }
+    return {};
+});
+
+let validateAd = (function(adItem){
+    if (typeof(adItem.id) == 'string' &&
+    typeof(adItem.description) == 'string' && adItem.description.length < 200  &&
+    adItem.createdAt instanceof Date &&  
+    typeof(adItem.link) == 'string' &&
+    typeof(adItem.vendor) == 'string' && adItem.vendor.length != 0 &&
+    adItem.hashTags instanceof Array &&
+    typeof(adItem.discount) == 'string' &&  adItem.discount.length != 0 &&
+    adItem.validUntil instanceof Date){
+        return true;
+    }else{
+        return false;
+    }
+});
+
+let addAd = (function(adItem){
+    if (adList.findIndex(ad => ad.id === adItem.id) >= 0) {
+        return false;
+    }
+   if (validateAd(adItem)) {
+        adList.push(adItem);
+         return true;
+    } else {
+        return false;
+    }
+});
+
+let editAd = (function(id,adItem){
+   let keys = Object.keys(adItem);
+    if(keys.includes('id') || keys.includes('author') || keys.includes('createdAt')){
+        return false;
+    }
+
+    let ind;
+    for(let i=0;i<=adList.length;++i){
+        if(i == adList.length){
+            return false;
+        }else if(adList[i].id === id){
+            ind = i;
+            break;
+        }
+    }
+    new_ad = adList[ind];
```

Comment:
You copied a link to old ad item here, not a value, then you changed original array item. So to make this function work properly, here you need to copy object itself (by using destructuring or Object.asign)